### DISCUSSION TITLE
Proxy WebSockets

### DISCUSSION THREAD

**Discussion by SilverRat** [chimurai/http-proxy-middleware#773] [Q&A]
I'm using http-proxy-middleware to direct students from a front end to a dedicated lab machine.  When they come to the lab page for the first time I assign them a unique cookie (which is just the private IP at this point in development) to identify them for routing later.  The initial setup seems to work fine for standard http traffic, but the product I'm working with appears to use WebSockets as well for at least one function and the cookie is not readable in the router function on that WebSocket call to the /api.  Any tips on how to set/capture something I can use for routing on the WebSocket call?  Thanks for helping.

```
app.use(['/auth','/admin','/api','/rules','/insights','/devops'], createProxyMiddleware({
    changeOrigin: false,  
    ws: true,  
    onProxyReq: fixRequestBody,   
    router: myRouter,
}));

function myRouter(req) {
    var labURL="";
    //Check for IP to route in cookie (cookie set when loading lab page)
    if (typeof req.cookies !== 'undefined' && req.cookies.labIP !== 'undefined') {
        labURL = req.cookies.labIP;
    } else {        
        // if IP for route not in cookie, we have a problem
        console.log('**Call to: ' + req.url + ' - Oh no, no cookie.  How can I route this?');
        //HACK:  HARD CODED IP to make routing work. MUST BE FIXED.
        labURL = 'http://44.210.215.200:8000';
    }
    return labURL;
}
```

**Comment by fonzane** [comment]
Did you find a solution for this issue? I'm facing the same problem trying to proxy node-red instances which run on IoT-Devices in a VPN and makes use of websockets. 🙄

**Reply by SilverRat** [reply]
I still haven't figured this one out.  Looking at it again now, but I'm wondering if this product just can't act as a proxy for WebSockets when you need to dynamically change the target based on the source. . .
